crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,725 @@
+    ])
+});
+
+#[pyclass(module = "qiskit._accelerate.commutation_checker")]
+struct CommutationChecker {
+    library: CommutationLibrary,
+    cache_max_entries: usize,
+    cache: HashMap<(String, String), CommutationCacheEntry>,
+    #[pyo3(get)]
+    current_cache_entries: usize,
+    #[pyo3(get)]
+    _cache_miss: usize,
+    #[pyo3(get)]
+    _cache_hit: usize,
+    #[pyo3(get)]
+    gates: Option<HashSet<String>>,
+}
+
+#[pymethods]
+impl CommutationChecker {
+    #[pyo3(signature = (standard_gate_commutations=None, cache_max_entries=1_000_000, gates=None))]
+    #[new]
+    fn py_new(
+        standard_gate_commutations: Option<Bound<PyAny>>, // Send a bound here
+        cache_max_entries: usize,
+        gates: Option<HashSet<String>>,
+    ) -> Self {
+        // Initialize sets before they are used in the commutation checker
+        Lazy::force(&SKIPPED_NAMES);
+        Lazy::force(&NO_CACHE_NAMES);
+        Lazy::force(&SUPPORTED_OP);
+
+        CommutationChecker {
+            library: CommutationLibrary::new(standard_gate_commutations),
+            cache: HashMap::with_capacity(cache_max_entries),
+            cache_max_entries,
+            current_cache_entries: 0,
+            _cache_miss: 0,
+            _cache_hit: 0,
+            gates,
+        }
+    }
+
+    #[pyo3(signature=(op1, op2, max_num_qubits=3))]
+    #[allow(clippy::too_many_arguments)]
+    fn commute_nodes(
+        &mut self,
+        py: Python,
+        op1: &DAGOpNode,
+        op2: &DAGOpNode,
+        max_num_qubits: u32,
+    ) -> PyResult<bool> {
+        let mut bq: BitData<Qubit> = BitData::new(py, "qubits".to_string());
+        let op1_bound_qubits = op1.instruction.qubits.bind(py);
+        let op2_bound_qubits = op2.instruction.qubits.bind(py);
+        op1_bound_qubits
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        op2_bound_qubits
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        let qargs1 = op1_bound_qubits
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0)
+            .collect::<Vec<_>>();
+        let qargs2 = op2_bound_qubits
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0)
+            .collect::<Vec<_>>();
+
+        let mut bc: BitData<Clbit> = BitData::new(py, "clbits".to_string());
+        let op1_bound_clbit = op1.instruction.clbits.bind(py);
+        let op2_bound_clbit = op2.instruction.clbits.bind(py);
+
+        op1_bound_clbit
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+
+        op2_bound_clbit
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+
+        let cargs1 = op1_bound_clbit
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0)
+            .collect::<Vec<_>>();
+        let cargs2 = op2_bound_clbit
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0)
+            .collect::<Vec<_>>();
```

Comment:
The thing is though that `clbits` doesn't contain `c_if`/`.condition`. `cargs` is used for `Measure`, or `ControlFlowOp`. This is part of why we'll be deprecating and removing `condition`/`c_if` because it doesn't fit well into the data model. I don't think there are any instructions which could commute if they have clbits set on the instruction.